Fixed a SEGFAULT in setPose reported in #18

mkoval opened the pull request:
Previously, calling setPose() on an interactive marker causes a SEGFAULT if applyChanges() was not called on the server at least once since the marker was created. I traced the actual SEGFAULT to the doSetPose function. The value of header passed from setPose() is invalid because, in this case, marker_context_it = marker_contexts_.end().

I added a check for this case and, if there is no marker is present, instead use the header from the pending update. I'm not 100% sure that this is the correct solution, so it would be good to get input from someone who is more familiar with the code base.

Commit:
Fixed a SEGFAULT in setPose reported in #18

Previously, calling setPose() on an interactive marker causes a SEGFAULT
if applyChanges() was not called on the server at least once since the
marker was created. I traced the actual SEGFAULT to the doSetPose
function. The value of header passed from setPose() is invalid because,
in this case, marker_context_it = marker_contexts_.end().

I added a check for this case and, if there is no marker is present,
instead use the header from the pending update.

wjwwood commented on src/interactive_marker_server.cpp:
```diff
@@ -225,10 +225,21 @@ bool InteractiveMarkerServer::setPose( const std::string &name, const geometry_m
+    }
+    else
+    {
+      return false; // this should never happen
```
If this _were_ to happen, would returning false be the right thing to do? Maybe an exception should be raised or an assertion should be made, in the unlikely event that it occurs, so that this case can be more easily discovered.

dgossow replied:
An assertion would probably be the right thing, since you really should never hit this code path.

wjwwood commented:
Thanks for contribution, this seems reasonable to me, I am going to enable travis CI on this repo, at which point I'll get you to rebase this (so it can be tested).

Is anyone else interested in reviewing the pull request?

Reference #18

wjwwood commented:
Ok, I merged the travis integration, @mkoval could you please rebase against indigo-devel so that travis-ci can test it?

dgossow commented:
This looks good to me. Would you also be so kind to add a test in https://github.com/mkoval/interactive_markers/blob/SetPoseSEGFAULT/src/test/server_test.cpp?

Commit:
Replaced silent failure with an assertion.

Commit:
Added a setPose unit test.

mkoval commented:
I replaced the `return false` with an assertion, added a unit test, and rebased on indigo. Unfortunately, I can't figure out how to successfully run the unit test in Catkin, so that is untested. [Here is the output](https://gist.github.com/mkoval/0cb6d7ea8cfa9f381f02) I get when I run `catkin_make run_tests_interactive_markers`. This also happens on `hydro-devel`, so I presume it's not related to my change.

dgossow commented:
Awesome, thanks a bunch for fixing this.

There's still a build failure in server_test.cpp:88, you need to get rid of the ASSERT_TRUE.

Commit:
Removed an invalid ASSERT_TRUE.

mkoval commented:
I fixed that bug and the Travis build succeeded. I would, however, still love to know what I am doing wrong when running the unit tests locally. The documentation for unit tests in Catkin is rather sparse.

dgossow commented:
This looks good to go.

Regarding the testing, I can't really tell what the problem is, haven't used any of that stuff in quite a while. Maybe @wjwwood can help.

wjwwood commented:
Ok, I'll go ahead and merge this then, thanks for the contribution!

As for testing in catkin:
- http://docs.ros.org/indigo/api/catkin/html/howto/format1/run_tests.html
- http://answers.ros.org/question/54051/how-do-i-run-unit-tests-for-a-catkin-package/

mkoval commented:
Is there any way you can backport this to the `hydro-devel` branch? We haven't upgraded to Indigo yet (because it requires that we also upgrade to 14.04), and it's quite a pain to have everyone who uses our package build `interactive_markers` from source.

wjwwood commented:
Sure, I'll also actually release this into indigo, as it is just on the indigo-devel branch right now.

wjwwood commented:
Hydro release: https://github.com/ros/rosdistro/pull/6563
